pkg/query/endpointset.go

```diff
@@ -0,0 +1,704 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package query
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/thanos-io/thanos/pkg/exemplars/exemplarspb"
+	"github.com/thanos-io/thanos/pkg/info/infopb"
+	"google.golang.org/grpc"
+
+	"github.com/thanos-io/thanos/pkg/component"
+	"github.com/thanos-io/thanos/pkg/metadata/metadatapb"
+	"github.com/thanos-io/thanos/pkg/rules/rulespb"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	"github.com/thanos-io/thanos/pkg/store"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+	"github.com/thanos-io/thanos/pkg/targets/targetspb"
+)
+
+const (
+	unhealthyEndpointMessage = "removing endpoint because it's unhealthy or does not exist"
+
+	// Default minimum and maximum time values used by Prometheus when they are not passed as query parameter.
+	MinTime = -9223309901257974
+	MaxTime = 9223309901257974
+)
+
+type EndpointSpec interface {
+	// Addr returns Thanos API Address for the endpoint spec. It is used as ID for endpoint.
+	Addr() string
+	// Metadata returns current labels, component type and min, max ranges for store.
+	// It can change for every call for this method.
+	// If metadata call fails we assume that store is no longer accessible and we should not use it.
+	// NOTE: It is implementation responsibility to retry until context timeout, but a caller responsibility to manage
+	// given store connection.
+	Metadata(ctx context.Context, client infopb.InfoClient) (*endpointMetadata, error)
+
+	// StrictStatic returns true if the endpoint has been statically defined and it is under a strict mode.
+	StrictStatic() bool
+}
+
+type grpcEndpointSpec struct {
+	addr         string
+	strictstatic bool
+}
+
+// NewGRPCEndpointSpec creates gRPC endpoint spec.
+// It uses InfoAPI to get Metadata.
+func NewGRPCEndpointSpec(addr string, strictstatic bool) EndpointSpec {
+	return &grpcEndpointSpec{addr: addr, strictstatic: strictstatic}
+}
+
+// StrictStatic returns true if the endpoint has been statically defined and it is under a strict mode.
+func (es *grpcEndpointSpec) StrictStatic() bool {
```

Comment:
> Are we sure we are not creating something for tests? I think I did that for storeset. Otherwise we can remove yea

Yes. We only have one implementation which is also used in tests.